frame/election-providers/src/two_phase/mod.rs

```diff
@@ -0,0 +1,1325 @@
+pub struct RoundSnapshot<A> {
+	/// All of the voters.
+	pub voters: Vec<(A, VoteWeight, Vec<A>)>,
+	/// All of the targets.
+	pub targets: Vec<A>,
+}
+
+/// Some metadata related to snapshot.
+///
+/// In this pallet, there are cases where we want to read the whole snapshot (voters, targets,
+/// desired), and cases that we are interested in just the length of these values. The former favors
+/// the snapshot to be stored in one struct (as it is now) while the latter prefers them to be
+/// separate to enable the use of `decode_len`. This approach is a middle ground, storing the
+/// snapshot as one struct, whilst storing the lengths separately.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct RoundSnapshotMetadata {
+	/// The length of voters.
+	voters_len: u32,
+	/// The length of targets.
+	targets_len: u32,
+}
+
+/// Internal errors of the pallet.
+///
+/// Note that this is different from [`pallet::Error`].
+#[derive(RuntimeDebug, Eq, PartialEq)]
+pub enum InternalError {
+	/// A feasibility error.
+	Feasibility(FeasibilityError),
+	/// An error in the on-chain fallback.
+	OnChainFallback(crate::onchain::Error),
+	/// An internal error in the NPoS elections crate.
+	NposElections(sp_npos_elections::Error),
+	/// Snapshot data was unavailable unexpectedly.
+	SnapshotUnAvailable,
+	/// Submitting a transaction to the pool failed.
+	///
+	/// This can only happen in the unsigned phase.
+	PoolSubmissionFailed,
+}
+
+impl From<crate::onchain::Error> for InternalError {
+	fn from(e: crate::onchain::Error) -> Self {
+		InternalError::OnChainFallback(e)
+	}
+}
+
+impl From<sp_npos_elections::Error> for InternalError {
+	fn from(e: sp_npos_elections::Error) -> Self {
+		InternalError::NposElections(e)
+	}
+}
+
+impl From<FeasibilityError> for InternalError {
+	fn from(e: FeasibilityError) -> Self {
+		InternalError::Feasibility(e)
+	}
+}
+
+/// Errors that can happen in the feasibility check.
+#[derive(RuntimeDebug, Eq, PartialEq)]
+pub enum FeasibilityError {
+	/// Wrong number of winners presented.
+	WrongWinnerCount,
+	/// The snapshot is not available.
+	///
+	/// This must be an internal error of the chain.
+	SnapshotUnavailable,
+	/// Internal error from the election crate.
+	NposElection(sp_npos_elections::Error),
+	/// A vote is invalid.
+	InvalidVote,
+	/// A voter is invalid.
+	InvalidVoter,
+	/// A winner is invalid.
+	InvalidWinner,
+	/// The given score was invalid.
+	InvalidScore,
+	/// An error from the data provider's feasibility check
+	DataProvider(&'static str),
+}
+
+impl From<sp_npos_elections::Error> for FeasibilityError {
+	fn from(e: sp_npos_elections::Error) -> Self {
+		FeasibilityError::NposElection(e)
+	}
+}
+
+pub use pallet::*;
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + SendTransactionTypes<Call<Self>>
```

Comment:
what's the `SendTransactionTypes<Call<Self>>` about?

Reply:
This is a new requirement for pallets that want to submit txs from their OCW. im-online and staking currently bound the same. I believe it was added here: https://github.com/paritytech/substrate/pull/5182